expo/expo
# fix(cli,create-expo): stop corrupting binaries when transforming templates, via .expo-rename config

# Why

Continuing on from https://github.com/expo/expo/pull/26741 where I merely kicked the can down the road, here I implement a proper lasting solution.

As explained in the previous PR, both `create-expo` and `@expo/cli` corrupt certain files when extracting templates due to how they rename "HelloWorld" to the user's specified app name during extraction. Specifically, this seems to cause binary files (e.g. executables) to double in size, presumably due to an accidental encoding change.

While select file types are spared from this process [based on their file extension](https://github.com/expo/expo/blob/ce141b51fb8aaaa8e11038dd997ab6a3f22f2429/packages/create-expo/src/createFileTransform.ts#L60-L73), it's not sustainable to guess every possible file extension that should be exempt (we're missing `.zip`, for one thing), and overlooks files that have no extension at all (like binaries).

So I thought that maybe we've been doing it the wrong way round. Instead of renaming *all but some* files, we should **explicitly select which ones get renamed**. This would be configurable, of course – we'd have sensible defaults that would avoid most devs ever having to think about the process, but we'd also support templates including their own config to handle any exotic cases.

Importance: this allows us to safely commit to templates the following files:
* prebuilt binaries like `.xcframework`s
* arbitrary compressed files like `.zip`, for reducing download times
* more media types (all kinds of videos)
* files that you actually want to write "HelloWorld" into, for some reason

… It also finally makes the renaming procedure atomic and easily testable (hence the many new tests).

# How



The feature is implemented in both `@expo/cli` and `create-expo`:

* I removed the file contents rewriting process from the tar extractor (see changes to `createFileTransform.ts`).
* I kept the file path rewriting process of the tar extractor, however, as that seemed to do more good than harm (but yes, maybe deserves a followup).
* I introduced two post-extraction steps:
  1. `getTemplateFilesToRenameAsync()`, which accepts a "rename config", or otherwise falls back to the CLI's default (which hopefully covers most cases in existence anyway) to determine which files should be renamed
  2. `renameTemplateAppNameAsync()`, which, for a given list of files, renames “HelloWorld” to the desired app name with appropriate sanitsation
* I added unit tests and E2E tests for each of these functions
* I worked out a default rename config (search for `defaultRenameConfig`) that seems to cover all template cases in the monorepo, though I may have possibly overlooked some cases in templates outside the monorepo.

## About the rename config

See the documentation in the updated `packages/create-expo/README.md`.

# Test Plan



I've added lots of automated tests to give proper confidence, but manual testing can be performed by building Create Expo and using them:

* Check out the branch of interest (either `main` or this PR, `@shirakaba/extract-binaries-safely`)
* Build Create Expo and create a template with it
  ```sh
  # cd to wherever you keep your Expo monorepo
  cd ~/expo
  
  # Build create-expo
  cd packages/create-expo
  yarn build
  
  # cd to some other work directory. We'll build a template
  # in there.
  cd ~/Downloads
  
  ~/expo/packages/create-expo/build/index.js
  # Make an app called "ByeWorld", then move it into a
  # `new` or `old` directory for easy comparison later
  ```
* Having generated a template using a build of Create Expo derived from both `main` and this PR's branch, you can finally compare the created templates to assess that there have been no regressions. If this command produces no output, then there's no difference (and thus no regression):
  ```sh
  diff -r --exclude=.git new/ByeWorld old/ByeWorld
  ```

![image](https://github.com/expo/expo/assets/14055146/4271aeb0-1b4d-4d7e-92e4-927bca060af4)


# Checklist


# Related

https://github.com/expo/expo/pull/27071
